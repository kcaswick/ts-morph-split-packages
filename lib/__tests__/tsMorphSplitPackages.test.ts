// X import assert from "assert";
import { createWriteStream } from "fs";
import madge from "madge";
import { promisify } from "util";
import { PackageMapping } from "../index";

describe("tsMorphSplitPackages basic tests", () => {
  const simpleMadgeConfigPath = "lib/__tests__/simpleMadgeTestData/PackageMap.json";
  const simpleMadgeDependenciesPath = "lib/__tests__/simpleMadgeTestData/selfMadge.json";

  it("generate simpleMadgeTestData", async () => {
    const m = await madge(
      [
        "lib/index.ts",
        "./dist/cjs/index.cjs",
        "dist/esm/index.mjs",
        "./dist/index.js",
        "./dist/index.d.ts",
        "./dist/bundle.d.ts",
      ],
      {
        baseDir: ".",
        fileExtensions: ["ts", "tsx", "js", "jsx"],
        tsConfig: "tsconfig.json",
      }
    );
    expect(m.warnings()).toMatchInlineSnapshot(`
      Object {
        "skipped": Array [],
      }
    `);

    const dependencies = m.obj();
    expect(dependencies).toBeDefined();

    const s = createWriteStream(simpleMadgeDependenciesPath, {
      flags: "w+",
    });
    try {
      const dependenciesJson = JSON.stringify(dependencies, undefined, 2);
      await promisify(
        // eslint-disable-next-line no-unused-vars
        s.write.bind(s) as (chunk: string, _?: (error: Error | null | undefined) => void) => boolean
      )(dependenciesJson);
    } catch (err) {
      if (err) {
        console.error(err);
      }
    } finally {
      s.close();
    }
  });

  function loadSimpleMadge() {
    const m = new PackageMapping();
    m.getPackageMap(simpleMadgeDependenciesPath, simpleMadgeConfigPath);
    return m;
  }

  it("test loading simpleMadgeTestData", () => {
    const m = loadSimpleMadge();
    expect(m).toMatchSnapshot();
  });

  it("test mapPackage", () => {
    const m = loadSimpleMadge();
    const result = m.mapPackage("dist/index.js");
    expect(result).toBeDefined();
    expect(result?.Path).toBe("src/index.js");
    expect(result).toMatchObject({
      Path: "src/index.js",
      Repo: "new",
      Package: "new",
    });
  });

  it.todo("test bad config filename");
  it.todo("test bad dependency filename");
});
