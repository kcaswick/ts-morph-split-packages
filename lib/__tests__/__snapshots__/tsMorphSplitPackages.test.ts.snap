// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tsMorphSplitPackages basic tests test loading simpleMadgeTestData 1`] = `
PackageMapping {
  "config": Config {
    "BranchPrefix": "split/",
    "OldPatterns": Map {
      "lib/mapping" => Object {
        "Package": "new",
        "Path": "src/mapping",
        "Repo": "new",
      },
      "lib/__tests__/test_" => Object {
        "Package": "new",
        "Path": "src/__tests__/test_",
        "Repo": "new",
      },
      "lib" => Object {
        "Order": 10,
        "Package": "N/A",
        "Path": "lib",
        "Repo": "ts-morph-split-packages",
      },
    },
  },
  "depMap": Array [
    Object {
      "Name": MapResult {
        "New": Location {
          "Order": 10,
          "Package": "N/A",
          "Path": "lib/__tests__/complexRepo.test.ts",
          "Repo": "ts-morph-split-packages",
          "toString": [Function],
        },
        "OldName": "lib/__tests__/complexRepo.test.ts",
        "isUnmapped": [Function],
        "toString": [Function],
      },
      "dependencyMap": Array [
        MapResult {
          "New": Location {
            "Order": 10,
            "Package": "N/A",
            "Path": "lib/index.ts",
            "Repo": "ts-morph-split-packages",
            "toString": [Function],
          },
          "OldName": "lib/index.ts",
          "isUnmapped": [Function],
          "toString": [Function],
        },
      ],
    },
    Object {
      "Name": MapResult {
        "New": Location {
          "Order": 10,
          "Package": "N/A",
          "Path": "lib/__tests__/git.test.ts",
          "Repo": "ts-morph-split-packages",
          "toString": [Function],
        },
        "OldName": "lib/__tests__/git.test.ts",
        "isUnmapped": [Function],
        "toString": [Function],
      },
      "dependencyMap": Array [
        MapResult {
          "New": Location {
            "Order": 0,
            "Package": "new",
            "Path": "src/__tests__/test_fixtures.ts",
            "Repo": "new",
            "toString": [Function],
          },
          "OldName": "lib/__tests__/test_fixtures.ts",
          "isUnmapped": [Function],
          "toString": [Function],
        },
        MapResult {
          "New": Location {
            "Order": 10,
            "Package": "N/A",
            "Path": "lib/git.ts",
            "Repo": "ts-morph-split-packages",
            "toString": [Function],
          },
          "OldName": "lib/git.ts",
          "isUnmapped": [Function],
          "toString": [Function],
        },
      ],
    },
    Object {
      "Name": MapResult {
        "New": Location {
          "Order": 0,
          "Package": "new",
          "Path": "src/__tests__/test_fixtures.ts",
          "Repo": "new",
          "toString": [Function],
        },
        "OldName": "lib/__tests__/test_fixtures.ts",
        "isUnmapped": [Function],
        "toString": [Function],
      },
      "dependencyMap": Array [],
    },
    Object {
      "Name": MapResult {
        "New": Location {
          "Order": 10,
          "Package": "N/A",
          "Path": "lib/__tests__/tsMorphSplitPackages.test.ts",
          "Repo": "ts-morph-split-packages",
          "toString": [Function],
        },
        "OldName": "lib/__tests__/tsMorphSplitPackages.test.ts",
        "isUnmapped": [Function],
        "toString": [Function],
      },
      "dependencyMap": Array [
        MapResult {
          "New": Location {
            "Order": 0,
            "Package": "new",
            "Path": "src/__tests__/test_fixtures.ts",
            "Repo": "new",
            "toString": [Function],
          },
          "OldName": "lib/__tests__/test_fixtures.ts",
          "isUnmapped": [Function],
          "toString": [Function],
        },
      ],
    },
    Object {
      "Name": MapResult {
        "New": Location {
          "Order": 10,
          "Package": "N/A",
          "Path": "lib/git.ts",
          "Repo": "ts-morph-split-packages",
          "toString": [Function],
        },
        "OldName": "lib/git.ts",
        "isUnmapped": [Function],
        "toString": [Function],
      },
      "dependencyMap": Array [
        MapResult {
          "New": Location {
            "Order": 0,
            "Package": "new",
            "Path": "src/mapping.ts",
            "Repo": "new",
            "toString": [Function],
          },
          "OldName": "lib/mapping.ts",
          "isUnmapped": [Function],
          "toString": [Function],
        },
      ],
    },
    Object {
      "Name": MapResult {
        "New": Location {
          "Order": 10,
          "Package": "N/A",
          "Path": "lib/index.ts",
          "Repo": "ts-morph-split-packages",
          "toString": [Function],
        },
        "OldName": "lib/index.ts",
        "isUnmapped": [Function],
        "toString": [Function],
      },
      "dependencyMap": Array [
        MapResult {
          "New": Location {
            "Order": 10,
            "Package": "N/A",
            "Path": "lib/git.ts",
            "Repo": "ts-morph-split-packages",
            "toString": [Function],
          },
          "OldName": "lib/git.ts",
          "isUnmapped": [Function],
          "toString": [Function],
        },
        MapResult {
          "New": Location {
            "Order": 0,
            "Package": "new",
            "Path": "src/mapping.ts",
            "Repo": "new",
            "toString": [Function],
          },
          "OldName": "lib/mapping.ts",
          "isUnmapped": [Function],
          "toString": [Function],
        },
      ],
    },
    Object {
      "Name": MapResult {
        "New": Location {
          "Order": 0,
          "Package": "new",
          "Path": "src/mapping.ts",
          "Repo": "new",
          "toString": [Function],
        },
        "OldName": "lib/mapping.ts",
        "isUnmapped": [Function],
        "toString": [Function],
      },
      "dependencyMap": Array [],
    },
  ],
  "invdepMap": Array [
    Object {
      "Name": MapResult {
        "New": Location {
          "Order": 10,
          "Package": "N/A",
          "Path": "lib/index.ts",
          "Repo": "ts-morph-split-packages",
          "toString": [Function],
        },
        "OldName": "lib/index.ts",
        "isUnmapped": [Function],
        "toString": [Function],
      },
      "dependentMap": Array [
        MapResult {
          "New": Location {
            "Order": 10,
            "Package": "N/A",
            "Path": "lib/__tests__/complexRepo.test.ts",
            "Repo": "ts-morph-split-packages",
            "toString": [Function],
          },
          "OldName": "lib/__tests__/complexRepo.test.ts",
          "isUnmapped": [Function],
          "toString": [Function],
        },
      ],
    },
    Object {
      "Name": MapResult {
        "New": Location {
          "Order": 0,
          "Package": "new",
          "Path": "src/__tests__/test_fixtures.ts",
          "Repo": "new",
          "toString": [Function],
        },
        "OldName": "lib/__tests__/test_fixtures.ts",
        "isUnmapped": [Function],
        "toString": [Function],
      },
      "dependentMap": Array [
        MapResult {
          "New": Location {
            "Order": 10,
            "Package": "N/A",
            "Path": "lib/__tests__/git.test.ts",
            "Repo": "ts-morph-split-packages",
            "toString": [Function],
          },
          "OldName": "lib/__tests__/git.test.ts",
          "isUnmapped": [Function],
          "toString": [Function],
        },
        MapResult {
          "New": Location {
            "Order": 10,
            "Package": "N/A",
            "Path": "lib/__tests__/tsMorphSplitPackages.test.ts",
            "Repo": "ts-morph-split-packages",
            "toString": [Function],
          },
          "OldName": "lib/__tests__/tsMorphSplitPackages.test.ts",
          "isUnmapped": [Function],
          "toString": [Function],
        },
      ],
    },
    Object {
      "Name": MapResult {
        "New": Location {
          "Order": 10,
          "Package": "N/A",
          "Path": "lib/git.ts",
          "Repo": "ts-morph-split-packages",
          "toString": [Function],
        },
        "OldName": "lib/git.ts",
        "isUnmapped": [Function],
        "toString": [Function],
      },
      "dependentMap": Array [
        MapResult {
          "New": Location {
            "Order": 10,
            "Package": "N/A",
            "Path": "lib/__tests__/git.test.ts",
            "Repo": "ts-morph-split-packages",
            "toString": [Function],
          },
          "OldName": "lib/__tests__/git.test.ts",
          "isUnmapped": [Function],
          "toString": [Function],
        },
        MapResult {
          "New": Location {
            "Order": 10,
            "Package": "N/A",
            "Path": "lib/index.ts",
            "Repo": "ts-morph-split-packages",
            "toString": [Function],
          },
          "OldName": "lib/index.ts",
          "isUnmapped": [Function],
          "toString": [Function],
        },
      ],
    },
    Object {
      "Name": MapResult {
        "New": Location {
          "Order": 0,
          "Package": "new",
          "Path": "src/mapping.ts",
          "Repo": "new",
          "toString": [Function],
        },
        "OldName": "lib/mapping.ts",
        "isUnmapped": [Function],
        "toString": [Function],
      },
      "dependentMap": Array [
        MapResult {
          "New": Location {
            "Order": 10,
            "Package": "N/A",
            "Path": "lib/git.ts",
            "Repo": "ts-morph-split-packages",
            "toString": [Function],
          },
          "OldName": "lib/git.ts",
          "isUnmapped": [Function],
          "toString": [Function],
        },
        MapResult {
          "New": Location {
            "Order": 10,
            "Package": "N/A",
            "Path": "lib/index.ts",
            "Repo": "ts-morph-split-packages",
            "toString": [Function],
          },
          "OldName": "lib/index.ts",
          "isUnmapped": [Function],
          "toString": [Function],
        },
      ],
    },
  ],
}
`;
